### ISSUE TITLE
Analyzer completion default timeout should be increased

### ISSUE THREAD

**Issue by jensjoha** [dart-lang/sdk#51065]
The default completion request timeout in Analyzer is 100 ms:

https://github.com/dart-lang/sdk/blob/abae3062396ebf5474d9f8d162bf023b40320cd5/pkg/analysis_server/lib/src/services/completion/dart/completion_manager.dart#L55

Via intellij, what I in practise see is that incomplete queries are followed up with the same query again up to (seemingly) 8 times.
For an example slow query (https://github.com/flutter/flutter-intellij/issues/6470) it does exactly this: Asks, the query times out after ~100 ms, maybe 10-20 ms passes (probably the plugin making a decision based on the result), the plugin asks again, the query times out after ~100 ms and so on and so forth until it's made the same query 8 times.
In total the whole thing has taken almost a second and all the user got was a incomplete results (8 incomplete results, technically, but still).

If I increase the timeout what happens - granted, on this query, other queries will obviously behave differently - is that it responds with a full result after ~400 ms, is only asked once and the user waits less than half the time for a good result.

In my opinion this is better and I think the timeout should be increased.

PS: The 400 ms is because the flutter intellij plugin asks for "server.log" stuff and is subsequently send ~1.7 MB of data per request. Without it (e.g. the dart intellij plugin) it takes ~150 ms.

**Comment by karvulf** [comment]
The increase worked for me very well. 
The only issue I have, that I cannot choose any devices on my IDE when changing the dart sdk path inside the flutter folder to my custom sdk.

**Comment by karvulf** [comment]
Unforntunately, it seemed to work but I couldn't solve the issue to build a custom dart-sdk version using flutter. The main issue is the Kernel SDK version check. I could remove it from the dart-sdk but flutter is also checking for the Kernel SDK version. So I hope this timeout can be increased for larger projects and will be added to a newer flutter version.

**Comment by srawlins** [comment]
cc @jwren @scheglov

**Comment by scheglov** [comment]
We mix two things here:
1. The default timeout.
2. The plugin behavior.

The default timeout is intentionally low, so that even if we don't have full element models for not yet imported libraries, we still quickly return results from local scopes, and imported libraries. They are ready, and probably will be returned fast. And when we stopped on time budget, the response says `"isIncomplete":true`.

But the way plugin reacts to this seems problematic to me. It re-requests completion suggestions, so abandons the idea "use whatever is available".

In case of Flutter, I see completion requested 5 times, but once all caches are warmed, subsequent results a returned in `50 ms`.

**Comment by jacob314** [comment]
Fyi @jwren for the question of whether we should change how the plugin reacts in this case.

**Comment by jensjoha** [comment]
Having a timeout - especially a low one - on completion means that people with slower machines get worse completion than people with fast machines. In my opinion that hardly seems ideal.

We can also have two similar requests where the results are different because one happened to finish within the timeout and one didn't. That must be very confusing for users.

I personally see no problem in increasing the timeout (significantly) --- the plugins seem to cancel the requests if not needed anyway (e.g. if the user keeps typing).

While I do get the point of returning local results fast, isn't the idea of not imported libraries not yet being ready only the case when we're starting up?

(ideally it seems like the call for completion should be two requests instead (or somehow answered as two) one with "fast" results and one with "slower" results).

**Comment by bwilkerson** [comment]
> (ideally it seems like the call for completion should be two requests instead (or somehow answered as two) one with "fast" results and one with "slower" results)

You might be referring to a style that we experimented with earlier in which results could be returned in batches so that slower results were returned after faster results (without requiring changes from the user), but it wasn't possible to implement that mode given the current behavior of our clients. We are constrained by the capabilities of the clients.

That said, the API does allow for something very close to this. Both LSP and the legacy protocol have a Boolean flag on the response that indicates whether the results are incomplete. An incomplete list of responses means that the list of suggestions will be re-requested as the user types.

> I personally see no problem in increasing the timeout (significantly) ...

The 100 ms timeout should (by which I mean ideally) almost never be hit. Delays of more than 100 ms significantly decrease the value of providing suggestions at all. While it might make sense temporarily to increase the limit, the limit, imo, isn't the problem. The problem is that we're taking long enough to compute suggestions that we ever got close to the limit, and that's the problem we need to be trying to solve.

**Comment by mraleph** [comment]
I agree with the premise that in general completions should be computed and returned quickly. That being said we should consider if we want to degrade _gracefully_ or not. The current degradation is not graceful - if for some reason completion takes too long to compute the UX gets really bad. If for some reason a regression sneaks out which makes completion queries exceed timeout then completion breaks down completely, instead of just slowing down. 

There is also another question I got when reading @jensjoha's description: why does repeating the same query consecutively 8 times hits timeout 8 times? I would have expected some sort of caching maybe especially when computations take so long and might timeout. So e.g. for completion query that takes 150ms the IDE should ask only two times: first will timeout, but the second one will succeed using the cached result?

**Comment by keertip** [comment]
Lowering priority as this, as it looks to be a longer arc of work, lof improving completion suggestion computations than fixed by increasing the timeout.
